docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+parallel `tables` module, with a nearly identical set of type definitions: for
+instance, [both crates][read-name-record] [contain a][write-name-record] `tables::name::NameRecord` type.
+
+We will examine each of these crates separately.
+
+## <a id="read-fonts"></a> `read-fonts`
+
+In the [`read-fonts`][] crate, we make a distinction between *table* objects and
+*record* objects, and we generate different code for each.
+
+The distinction between a *table* and a *record* is blurry, but the
+specification offers two "general criteria":
+
+> - Tables are referenced by offsets. If a table contains an offset to a
+> sub-structure, the offset is normally from the start of that table.
+> - Records occur sequentially within a parent structure, either within a
+> sequence of table fields or within an array of records of a given type. If a
+> record contains an offset to a sub-structure, that structure is logically a
+> subtable of the record’s parent table and the offset is normally from the start
+> of the parent table.
+>
+> ([The OpenType font file][otff])
+
+### <a id="scalars-detour"></a> A brief detour on scalars and `BigEndian<T>`
+
+#### a description of the problem
+
+Before we dive into the specifics of the tables and records in `read-fonts`, I
+want to talk briefly about how we represent and handle the [basic data types](ot-data-types)
+of which records and tables are composed.
+
+In the font file, these values are all represented in [big-endian][endianness]
+byte order. When we access them, we will need to convert them to the native
+endianness of the host platform. We also need to have some set of types which
+exactly match the memory layout (including byte ordering) of the underlying font
+file; this is necessary for us to take advance of zerocopy semantics (see the
+[zerocopy section](#zerocopy) below.)
+
+This leads us to a situation where we require two distinct types for each
+scalar: a native type that we will use in our program logic, and a
+'raw' type that will represent the bytes in the font file (as well as some
+mechanism to convert between them.)
+
+There are various ways we could express this in Rust. The most straightforward
+would be to just have two parallel sets of types: for instance alongside the
+`F2Dot14` type, we might have `RawF2Dot14`, or `F2Dot14Be`. Another option might
+be to have types that are generic over byte-order, such that you end up with
+types like `U16<BE>` and `U16<LE>`.
+
+There's one additional complication around these values: certain types in the
+spec are stored as three bytes, (`uint24` and `Offset24`) but in native code it
+is easier and more efficient to represent these as 32-bit integers. This means
+that the raw and normal types may not have the same size.
+
+I have taken a slightly different approach, which tries to be more ergonomic and
+intuitive to the user, at the cost of having a slightly more complicated
+implementation.
```

Comment:
IMHO you could delete "at the cost of having a slightly more complicated implementation" since you directly discuss the implementation below.

Reply:
I suggest it would be helpful to compare/contrast with HarfBuzz.

Reply:
I'm not sure I know enough about harfbuzz? My understanding is that harfbuzz takes advantage of C++'s more liberal overloading to perform conversion when the value is accessed/read, but I'm not sure if there's anything else interesting going on?

Reply:
To an extent I see making the explicit refs as a means to credit the giants whose shoulders we stand on :)